src/term.rs

```diff
@@ -468,6 +483,41 @@ pub mod record {
             RecordData { fields, ..self }
         }
     }
+
+    /// The sealed tail of a Nickel record under a polymorphic contract.
+    ///
+    /// Note that access to the enclosed [term] must only be allowed when a
+    /// matching [sealing_key] is provided. If this is not enforced it will
+    /// lead to parametricity violations.
+    #[derive(Clone, Debug, PartialEq)]
+    pub struct SealedTail {
+        /// The key with which the tail is sealed.
+        sealing_key: SealingKey,
+        /// The label to which blame will be attributed if code tries to
+        /// interact with the sealed tail in any way.
+        pub label: Label,
+        /// The term which is sealed.
+        term: RichTerm,
```

Comment:
Does it make sense to seal something else than a Record? We are evaluating the argument anyway I think (because this is an argument of a primop). The advantage of restricting what can appear here is that we could make better error messages, like instead of this field doesn't exist, we could say "oh, you're trying to access something in the tail, but the type of your function doesn't allow you to do so, you should do that and that".

Reply:
Yep, completely agree with this. I'd originally planned to make this change when we actually implement the improved error messages, but that's likely to just cause confusion if it's not me that ends up doing it. Will fix this before I merge.

Reply:
Actually, after investigating this, I don't think we should hold on to the `RecordData` here directly. The key issue is that in `SealRecordTail` we need to closurize the entire record's `RichTerm`. We could instead closurize the values of the record, but unless I'm mistaken we'd have to pay the cost of cloning the tail record's env for each entry.

As an alternative, we can take the record argument we receive in `SealRecordTail` and just build a `HashSet` of its keys which we can then check against whenever we would normally raise a missing field error. So, unless I've misunderstood something, I'll leave this as-is & just add a note to #952 that this might be the easiest way to implement this.

Reply:
Ah, it's true that the closurization is slightly more involved.

> We could instead closurize the values of the record, but unless I'm mistaken we'd have to pay the cost of cloning the tail record's env for each entry.

That being said:

1. Cloning environment is very cheap. Environments are `Rc`-ed, so it's all in all an `Rc::clone`.
2. Record fields must already be closurized. If you re-closurize something already closurized, there is an optimization which avoids adding a useless indirection. So mostly, closurizing would take the Closure out of the environment given as an argument, and put it in the outer environment. We may have to clone an environment to immediately drop it, which is not optimal, but should be pretty insignificant overall.

But the picture is less clear-cut. I think both approaches are valid. Yours may actually be easier. 

Though I think the seal record tail primitive should still fail if it doesn't see a record, even if you keep the current representation. Currently you're taking the tail out and merging it inside the implementation of the record tail contract. If a user (or our faulty implementation) injects something else in the tail, this going to make for bad error messages.

If we do this, you keep your easy closurizing, but we just ensure through sealing that we only actually put a closurized record inside. Doing so, we can very much get back this record from the error reporting code in fact: just get the thunk from the environment and inspect the closure, which must be an evaluated record.

Reply:
> Though I think the seal record tail primitive should still fail if it doesn't see a record, even if you keep the current representation.

Yep, totally agree. Will fix this now. 

Good point about the clones being cheaper than I was thinking. I'll defer the decision until we actually implement the field access blame error, and then we can see what makes most sense at the time. (Though tbh implementing that error feels pretty trivial whatever we do, so I might just do it quickly this week if I find some time. :grin:)